cpp-package/src/inc/brainflow_array.h

```diff
@@ -0,0 +1,468 @@
+    }
+
+    BrainFlowArray (int size0, int size1)
+        : length (size0 * size1)
+        , size (make_array (size0, size1))
+        , stride (make_stride<2> (make_array (size0, size1)))
+        , origin (nullptr)
+    {
+        static_assert (Dim == 2, "This function is only for BrainFlowArray<T, 2>");
+        origin = new T[length];
+        memset (origin, 0, sizeof (T) * length);
+    }
+
+    BrainFlowArray (int size0, int size1, int size2)
+        : length (size0 * size1 * size2)
+        , size (make_array (size0, size1, size2))
+        , stride (make_stride<3> (make_array (size0, size1, size2)))
+        , origin (nullptr)
+    {
+        static_assert (Dim == 3, "This function is only for BrainFlowArray<T, 3>");
+        origin = new T[length];
+        memset (origin, 0, sizeof (T) * length);
+    }
+
+    BrainFlowArray (T *ptr, int size0)
+        : length (size0)
+        , size (make_array (size0))
+        , stride (make_stride<1> (make_array (size0)))
+        , origin (nullptr)
+    {
+        static_assert (Dim == 1, "This function is only for BrainFlowArray<T, 1>");
+        origin = new T[size0];
+        memcpy (origin, ptr, size0 * sizeof (T));
+    }
+
+    BrainFlowArray (T *ptr, int size0, int size1)
+        : length (size0 * size1)
+        , size (make_array (size0, size1))
+        , stride (make_stride<2> (make_array (size0, size1)))
+        , origin (nullptr)
+    {
+        static_assert (Dim == 2, "This function is only for BrainFlowArray<T, 2>");
+        origin = new T[size0 * size1];
+        memcpy (origin, ptr, size0 * size1 * sizeof (T));
+    }
+
+    BrainFlowArray (T *ptr, int size0, int size1, int size2)
+        : length (size0 * size1 * size2)
+        , size (make_array (size0, size1, size2))
+        , stride (make_stride<3> (make_array (size0, size1, size2)))
+        , origin (nullptr)
+    {
+        static_assert (Dim == 3, "This function is only for BrainFlowArray<T, 3>");
+        origin = new T[size0 * size1 * size2];
+        memcpy (origin, ptr, size0 * size1 * size2 * sizeof (T));
+    }
+
+    explicit BrainFlowArray (const BrainFlowArray &other)
+        : length (other.length), size (other.size), stride (other.stride), origin (nullptr)
+    {
+        origin = new T[other.length];
+        memcpy (origin, other.origin, sizeof (T) * other.length);
+    }
+
+    const BrainFlowArray &operator= (const BrainFlowArray &other)
+    {
+        length = other.length;
+        size = other.size;
+        stride = other.stride;
+        origin = new T[other.length];
+        memcpy (origin, other.origin, sizeof (T) * other.length);
+
+        return *this;
+    }
+
+    BrainFlowArray (BrainFlowArray &&other)
+        : length (other.length)
+        , size (std::move (other.size))
+        , stride (std::move (other.stride))
+        , origin (nullptr)
+    {
+        origin = new T[other.length];
+        memcpy (origin, other.origin, sizeof (T) * other.length);
```

Comment:
Here there is no need for a memcpy.  You can simply say
```
origin = other.origin;
other.origin = nullptr;
```

Reply:
If you do keep the memcpy, you'll want to `delete [] other.origin` here to prevent a memory leak.